src/Storages/MergeTree/MergeTreeWhereOptimizer.cpp

```diff
@@ -225,7 +225,15 @@ void MergeTreeWhereOptimizer::optimize(ASTSelectQuery & select) const
             break;
 
         /// 10% ratio is just a guess.
-        if (total_size_of_moved_conditions > 0 && (total_size_of_moved_conditions + it->columns_size) * 10 > total_size_of_queried_columns)
+        /// If sizes of compressed columns cannot be calculated, e.g. for compact parts,
+        /// use number of moved columns as a fallback.
+        bool moved_enough =
```

Comment:
What does the size of moved condition mean? Do we try to move conditions on narrow or on wide columns firstly? Also I don't understand the comparator of Condition class :( I mean `viable` and `good` fields.

Reply:
The size of condition is the sum of compressed sizes of columns, that are in the condition. `viable` and `good` are just some heuristics on how selective predicate is. Firstly we try to move conditions with high selectivity and small size of required columns to avoid reading much data.

Reply:
Ok, I will try to add more columns.